README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# aeddo <img src="man/figures/logo.png" align="right" height="139" alt="" />

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/aeddo)](https://CRAN.R-project.org/package=aeddo)
[![R-CMD-check](https://github.com/ssi-dk/aeddo/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/ssi-dk/aeddo/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/ssi-dk/aeddo/branch/main/graph/badge.svg)](https://app.codecov.io/gh/ssi-dk/aeddo?branch=main)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
<!-- badges: end -->

## aeddo: Automated and Early Detection of Disease Outbreaks

aeddo is an R package that provides a cutting-edge solution for the
automated and early detection of disease outbreaks in time series data.
This innovative tool leverages hierarchical models in a novel way to
infer one-step ahead random effects, which are subsequently used to
identify and characterize disease outbreaks.

## Why aeddo?

Epidemiologists, public health professionals, and researchers are often
challenged by the need to detect disease outbreaks promptly. Timely
identification is critical for implementing appropriate control measures
and mitigating the impact of outbreaks. aeddo is designed to address
this challenge and offer a range of benefits:

- **Innovative Approach**: aeddo employs hierarchical models to estimate
  one-step ahead random effects, enabling early detection of outbreaks
  in disease data. This approach is a game-changer in epidemiological
  analysis.
- **Effectively Controls False Alarms**: aeddo’s applications are broad
  and relevant across the field of epidemiology. It supports
  professionals and researchers in various domains, including public
  health, epidemiological research, and data analysis, while effectively
  controlling the number of false alarms.
- **Real-time Insights**: By providing prospective detection
  capabilities, aeddo offers real-time insights into disease outbreak
  patterns. This is invaluable for making data-driven decisions and
  implementing timely interventions.

## Installation

``` r
# Install aeddo from CRAN
install.packages("aeddo")
```

### Development version

You can install the development version of `aeddo` from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("ssi-dk/aeddo")
```

Start leveraging aeddo today to gain a unique perspective on disease
outbreak detection in your time series data.

## Get Started

Explore the package’s functions and features by referring to the
documentation. Detailed examples and use cases are provided to help you
make the most of aeddo in your epidemiological work.

## Contributions

Contributions to aeddo are welcomed. If you have suggestions, feature
requests, or encounter issues, please don’t hesitate to open an
[issue](https://github.com/ssi-dk/aeddo/issues) or submit a [pull
request](https://github.com/ssi-dk/aeddo/pulls).


NEWS.md
# aeddo (development version)

## Bug fix

* Wrap the dispersion variable `phi` in `exp()` to ensure that it is a possible value.

## Improvements

* Now the starting guess for the models parameters, `theta`, is based on the previous round of optimization. This increases the performance and robustness of the paramter estimation.

# aeddo 0.1.1

## Patch

* Transferring maintainership of the R package to Lasse Engbo Christiansen.

# aeddo 0.1.0

## Features

- Added a new function `nll_poisson_gamma()` for computing the negative log-likelihood of a Poisson-Gamma model. This function is essential for parameter estimation in the `aeddo()` algorithm.
- Introducing the `aeddo()` function for Automated and Early Detection of Disease Outbreaks. This innovative algorithm utilizes hierarchical models to infer one-step ahead random effects, providing an effective tool for identifying and responding to disease outbreaks in time series data. The function includes various parameters for customization, such as rolling window size, outbreak detection threshold, and optimization method.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(aeddo)

test_check("aeddo")


tests/testthat/test-aeddo.R
test_that("Proper 'aedseo' class is returned", {
  # Sample some data
  tbl_data <- tibble::tibble(
    time = 1:100,
    y = rnbinom(
      n = 100,
      size = 10,
      prob = 0.7
    ),
    n = 1
  )

  # Run the algorithm
  aeddo_results <- aeddo(
    data = tbl_data,
    formula = y ~ 1,
    k = 36,
    sig_level = 0.95,
    exclude_past_outbreaks = TRUE,
    init_theta = c(1, 1),
    lower = c(-Inf, 1e-6),
    upper = c(Inf, 1e2),
    method = "L-BFGS-B"
  )

  # Test that the correct class is returned
  expect_s3_class(aeddo_results, "aeddo")
})


tests/testthat/test-internal.R
test_that("check_nll_poisson_gamma_inputs throws an error for invalid inputs", {
  # Invalid inputs for testing
  invalid_theta <- "invalid"
  invalid_data <- data.frame(x = c(1, 2, 3), n = c(7, 8, 9))
  invalid_data_class <- list(x = c(1, 2, 3), y = c(1, 2, 3), n = c(7, 8, 9))
  invalid_formula <- "invalid"
  valid_formula <- y ~ log(n)

  # Test invalid theta
  expect_error(
    check_nll_poisson_gamma_inputs(invalid_theta, invalid_data, valid_formula)
  )

  # Test missing 'y' in data
  expect_error(
    check_nll_poisson_gamma_inputs(1:3, invalid_data, valid_formula)
  )

  # Test invalid data class
  expect_error(
    check_nll_poisson_gamma_inputs(1:3, invalid_data_class, valid_formula)
  )

  # Test invalid formula
  expect_error(
    check_nll_poisson_gamma_inputs(
      1:3,
      data.frame(y = c(4, 5, 6), n = c(7, 8, 9)),
      invalid_formula
    )
  )
})

test_that(
  "check_nll_poisson_gamma_inputs does not throw an error for valid inputs",
  {
    # Valid inputs for testing
    valid_theta <- c(1, 2, 3)
    valid_data <- data.frame(y = c(4, 5, 6), n = c(7, 8, 9))
    valid_formula <- ~ y + n

    # Test valid inputs
    expect_no_error(
      check_nll_poisson_gamma_inputs(valid_theta, valid_data, valid_formula)
    )
  }
)

test_that("check_aeddo_inputs throws an error for invalid inputs", {
  # Invalid inputs for testing
  invalid_data <- data.frame(x = c(1, 2, 3), n = c(7, 8, 9))
  invalid_data_class <- list(x = c(1, 2, 3), y = c(1, 2, 3), n = c(7, 8, 9))
  invalid_formula <- "not_formula"
  invalid_k <- -5
  invalid_sig_level <- 1.5
  invalid_exclude_past_outbreaks <- "not_logical"
  invalid_init_theta <- "not_numeric"
  invalid_lower <- "not_numeric"
  invalid_upper <- "not_numeric"
  invalid_method <- "invalid_choice"
  valid_data <- data.frame(time = 1:3, y = c(4, 5, 6), n = c(7, 8, 9))
  valid_formula <- y ~ 1
  valid_k <- 2
  valid_sig_level <- 0.95
  valid_exclude_past_outbreaks <- TRUE
  valid_init_theta <- 1
  valid_lower <- 0
  valid_upper <- 100
  valid_method <- "BFGS"

  # Test missing 'y' in data
  expect_error(
    check_aeddo_inputs(
      invalid_data,
      valid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    )
  )

  # Test invalid data class
  expect_error(
    check_aeddo_inputs(
      invalid_data_class,
      valid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    )
  )

  # Test invalid formula
  expect_error(
    check_aeddo_inputs(
      valid_data,
      invalid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    )
  )

  # Test invalid k
  expect_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      invalid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    )
  )

  # Test invalid sig_level
  expect_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      valid_k,
      invalid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    ),
  )

  # Test invalid exclude_past_outbreaks
  expect_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      valid_k,
      valid_sig_level,
      invalid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    )
  )

  # Test invalid init_theta
  expect_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      invalid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    )
  )

  # Test invalid lower
  expect_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      invalid_lower,
      valid_upper,
      valid_method
    )
  )

  # Test invalid upper
  expect_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      invalid_upper,
      valid_method
    )
  )

  # Test invalid method
  expect_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      invalid_method
    )
  )
})

test_that("check_aeddo_inputs does not throw an error for valid inputs", {
  # Valid inputs for testing
  valid_data <- data.frame(time = 1:3, y = c(4, 5, 6), n = c(7, 8, 9))
  valid_formula <- y ~ 1
  valid_k <- 2
  valid_sig_level <- 0.95
  valid_exclude_past_outbreaks <- TRUE
  valid_init_theta <- 1
  valid_lower <- 0
  valid_upper <- 100
  valid_method <- "BFGS"

  # Test valid inputs
  expect_no_error(
    check_aeddo_inputs(
      valid_data,
      valid_formula,
      valid_k,
      valid_sig_level,
      valid_exclude_past_outbreaks,
      valid_init_theta,
      valid_lower,
      valid_upper,
      valid_method
    )
  )
})


R/plot.R
#' Create a complete 'ggplot' appropriate to a particular data type
#'
#' @description
#' `r lifecycle::badge('stable')`
#'
#'  This function generates a complete 'ggplot' object suitable for
#'  visualizing time series data in an `aeddo` object. It creates a line
#'  plot connecting the observations and adds points at each data point.
#'
#' @param x An `aeddo` object
#' @param ... Additional arguments (not used).
#'
#' @return A 'ggplot' object for visualizing the time series data.
#'
#' @aliases plot
#'
#' @seealso [autoplot()]
#'
#' @examples
#' # Create an example aeddo object
#' aeddo_data <- data.frame(
#'   time = as.Date(c(
#'     "2023-01-01",
#'     "2023-01-02",
#'     "2023-01-03",
#'     "2023-01-04",
#'     "2023-01-05",
#'     "2023-01-06"
#'   )),
#'   y = c(100, 120, 180, 110, 130, 140),
#'   n = 1
#' )
#'
#' # Supply a model formula
#' fixed_effects_formula <- y ~ 1
#'
#' # Choose a size for the rolling window
#' k <- 2
#' # ... and quantile for the threshold
#' sig_level <- 0.9
#'
#' # Employ the algorithm
#' aeddo_results <- aeddo(
#'   data = aeddo_data,
#'   formula = fixed_effects_formula,
#'   k = k,
#'   sig_level = sig_level,
#'   exclude_past_outbreaks = TRUE,
#'   init_theta = c(1, 0),
#'   lower = c(-Inf, 1e-6),
#'   upper = c(Inf, 1e2),
#'   method = "L-BFGS-B"
#' )
#'
#' # Create a ggplot visualization for the aeddo object
#' plot(aeddo_results)
#' @importFrom graphics plot
#' @rdname plot
#' @method plot aeddo
#' @export
plot.aeddo <- function(x, ...) {
  suppressWarnings(print(autoplot(x, ...)))
}


R/autoplot.R
#' Create a complete 'ggplot' appropriate to a particular data type
#'
#' @description
#' `r lifecycle::badge('stable')`
#'
#'  This function generates a complete 'ggplot' object suitable for
#'  visualizing time series data in an `aeddo` object. It creates a line
#'  plot connecting the observations and adds points at each data point.
#'
#' @param object An `aeddo` object
#' @param ... Additional arguments (not used).
#'
#' @return A 'ggplot' object for visualizing the time series data.
#'
#' @aliases autoplot
#'
#' @examples
#' # Create an example aeddo object
#' aeddo_data <- data.frame(
#'   time = as.Date(c(
#'     "2023-01-01",
#'     "2023-01-02",
#'     "2023-01-03",
#'     "2023-01-04",
#'     "2023-01-05",
#'     "2023-01-06"
#'   )),
#'   y = c(100, 120, 180, 110, 130, 140),
#'   n = 1
#' )
#'
#' # Supply a model formula
#' fixed_effects_formula <- y ~ 1
#'
#' # Choose a size for the rolling window
#' k <- 2
#' # ... and quantile for the threshold
#' sig_level <- 0.9
#'
#' # Employ the algorithm
#' aeddo_results <- aeddo(
#'   data = aeddo_data,
#'   formula = fixed_effects_formula,
#'   k = k,
#'   sig_level = sig_level,
#'   exclude_past_outbreaks = TRUE,
#'   init_theta = c(1, 0),
#'   lower = c(-Inf, 1e-6),
#'   upper = c(Inf, 1e2),
#'   method = "L-BFGS-B"
#' )
#'
#' # Create a ggplot visualization for the aeddo object
#' autoplot(aeddo_results)
#' @importFrom ggplot2 autoplot
#' @importFrom rlang .data
#' @rdname autoplot
#' @export
autoplot <- function(object, ...) {
  UseMethod("autoplot")
}
#' @rdname autoplot
#' @method autoplot aeddo
#' @export
autoplot.aeddo <- function(object, ...) {
  # Extract the supplied data
  data <- attr(object, "data")
  # ... and the window width
  k <- attr(object, "k")

  # Unnest the results
  unnested_object_results <- object %>%
    tidyr::unnest("reference_data")

  # Join with original data, to visualize full series
  joined_results <- data %>%
    dplyr::full_join(
      y = unnested_object_results,
      by = dplyr::join_by("time", "y", "n")
    )

  # Extract the observations used for training
  training_dates <- utils::head(data$time, k)

  # Make a nice 'ggplot' visualization
  suppressWarnings(
    joined_results %>%
      ggplot2::ggplot(
        mapping = ggplot2::aes(
          x = .data$time,
          y = .data$y,
          alpha = .data$outbreak_alarm
        )
      ) +
      ggplot2::geom_point() +
      ggplot2::annotate(
        geom = "rect",
        xmin = training_dates[1] - Inf,
        xmax = training_dates[k],
        ymin = -Inf,
        ymax = Inf,
        alpha = 0.2
      ) +
      ggplot2::guides(alpha = "none")
  )
}


R/internal.R
##' Check and assert the inputs for the aeddo function.
#'
#' This function is for internal use and checks the validity of inputs to the
#' aeddo function. It performs various checks using the checkmate package to
#' ensure that the inputs conform to expected types and values.
#'
#' @param data A data frame containing the time series data.
#' @param formula A model formula for the fixed effects in the hierarchical
#' model.
#' @param k An integer specifying the rolling window size employed for
#' parameter estimation.
#' @param sig_level The quantile from the random effects distribution used for
#' defining the outbreak detection threshold.
#' @param exclude_past_outbreaks Logical indicating whether past
#' outbreak-related observations should be excluded from future parameter
#' estimation.
#' @param init_theta Initial values for model parameters in optimization.
#' @param lower Lower bounds for optimization parameters.
#' @param upper Upper bounds for optimization parameters.
#' @param method The optimization method to use, either "BFGS" or "L-BFGS-B".
#'
#' @return Returns nothing. Raises errors if the inputs do not meet the
#' specified criteria.
#'
#' @keywords internal
#'
#' @examples
#' \dontrun{
#' # This function is for internal use and is called within the aeddo function.
#' # It is not intended to be called directly by users.
#' }
check_aeddo_inputs <- function(
    data,
    formula,
    k,
    sig_level,
    exclude_past_outbreaks,
    init_theta,
    lower,
    upper,
    method) {
  coll <- checkmate::makeAssertCollection()

  checkmate::assert_class(
    data,
    classes = c("data.frame"),
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_names(
    colnames(data),
    must.include = c("time", "y", "n"),
    add = coll
  )
  checkmate::assert_formula(
    formula,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_count(
    k,
    na.ok = FALSE,
    positive = TRUE,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_number(
    sig_level,
    na.ok = FALSE,
    lower = 0,
    upper = 1,
    finite = TRUE,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_flag(
    exclude_past_outbreaks,
    na.ok = FALSE,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_numeric(
    lower,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_numeric(
    upper,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_numeric(
    init_theta,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_choice(
    method,
    choices = c("BFGS", "L-BFGS-B"),
    add = coll
  )

  # Report assertions
  checkmate::reportAssertions(coll)
}

#' Check and assert the inputs for the nll_poisson_gamma function.
#'
#' This function is for internal use and checks the validity of inputs to the
#' nll_poisson_gamma function. It performs various checks using the checkmate
#' package to ensure that the inputs conform to expected types and values.
#'
#' @param theta Numeric vector of parameters.
#' @param data A data frame containing the time series data.
#' @param formula A model formula for the fixed effects in the hierarchical
#' model.
#'
#' @return Returns nothing. Raises errors if the inputs do not meet the
#' specified criteria.
#'
#' @keywords internal
#'
#' @examples
#' \dontrun{
#' # This function is for internal use and is called within the
#' # nll_poisson_gamma function. It is not intended to be called directly by
#' # users.
#' }
check_nll_poisson_gamma_inputs <- function(
    theta,
    data,
    formula) {
  # Initialize the assertion collections
  coll <- checkmate::makeAssertCollection()

  # Assert function inputs
  checkmate::assert_numeric(
    theta,
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_class(
    data,
    classes = c("data.frame"),
    null.ok = FALSE,
    add = coll
  )
  checkmate::assert_names(
    colnames(data),
    must.include = c("y", "n"),
    add = coll
  )
  checkmate::assert_formula(
    formula,
    null.ok = FALSE,
    add = coll
  )

  # Report assertions
  checkmate::reportAssertions(coll)
}


R/poisson_gamma.R
poisson_gamma <- function(
    theta,
    y,
    n = 1,
    formula) {
  # Create a data.frame with the observation 'y' and population sizes 'n'
  observations_data <- data.frame(
    y = y,
    n = n
  )

  # Construct design matrix for the model
  design_matrix <- model.matrix(
    object = formula,
    data = observations_data
  )

  # Extract number of parameters in the fixed effects model
  n_parameters <- ncol(design_matrix)

  # Construct vector with fixed effects parameters
  beta <- theta[1:n_parameters]
  # ... and model parameters
  phi <- theta[-(1:n_parameters)]

  # Define the lambda parameter
  lambda <- exp(design_matrix %*% beta - log(n))

  # Construct "size" and "probability" for the negative binomial distribution
  size <- 1 / phi
  probability <- 1 / (lambda * phi + 1)

  # Calculate the log-likelihood
  log_likelihood <- sum(
    dnbinom(
      x = y,
      size = size,
      prob = probability,
      log = TRUE
    )
  )

  # Return the negative log likelihood
  negative_log_likelihood <- -log_likelihood

  return(negative_log_likelihood)
}


R/nll_poisson_gamma.R
#' Negative Log-Likelihood for Poisson Gamma Model
#'
#' @description
#' `r lifecycle::badge('stable')`
#'
#' Calculate the negative log-likelihood for the Poisson Gamma modeling
#' framework.
#'
#' @param theta A numeric vector containing model parameters. The first part of
#' the vector represents fixed effects, and the remaining part represents model
#' parameters.
#' @param data A tibble containing the time series data, including columns 'y'
#' for observed values,'n' for population size, and other covariates of
#' interest.
#' @param formula A formula specifying the model structure.
#'
#' @return The negative log-likelihood value.
#'
#' @export
#'
#' @examples
#' # Initial parameters
#' theta <- c(0.5, 0.1)
#'
#' # Sample data
#' data <- data.frame(
#'   y = c(10, 15, 20, 30, 50, 100, 200, 40, 20, 10),
#'   n = c(100, 150, 200, 300, 500, 1000, 2000, 400, 200, 100)
#' )
#'
#' # Fixed effects model formula
#' fixed_effects_formula <- y ~ 1
#'
#' # Calculate negative log likelihood
#' nll_poisson_gamma(
#'   theta = theta,
#'   data = data,
#'   formula = fixed_effects_formula
#' )
nll_poisson_gamma <- function(
    theta,
    data,
    formula) {
  # Assert function inputs
  check_nll_poisson_gamma_inputs(
    theta,
    data,
    formula
  )

  # Extract the observations
  y <- data$y
  # ... and population size
  n <- data$n

  # Construct design matrix for the model
  design_matrix <- stats::model.matrix(
    object = formula,
    data = data
  )

  # Extract number of parameters in the fixed effects model
  n_parameters <- ncol(design_matrix)

  # Construct vector with fixed effects parameters
  beta <- theta[1:n_parameters]
  # ... and model parameters
  phi <- exp(theta[-(1:n_parameters)])

  # Define the lambda parameter
  lambda <- exp(design_matrix %*% beta - log(n))

  # Construct "size" and "probability" for the negative binomial distribution
  size <- 1 / phi
  probability <- 1 / (lambda * phi + 1)

  # Calculate the log-likelihood
  log_likelihood <- sum(
    stats::dnbinom(
      x = y,
      size = size,
      prob = probability,
      log = TRUE
    )
  )

  # Return the negative log likelihood
  negative_log_likelihood <- -log_likelihood

  return(negative_log_likelihood)
}


R/aeddo.R
#' Automated and Early Detection of Disease Outbreaks
#'
#' @description
#' `r lifecycle::badge('stable')`
#'
#' This function performs automated an early detection of disease outbreaks,
#' (aeddo), on a time series data set. It utilizes hierarchical models in an
#' innovative manner to infer one-step ahead random effects. In turn, these
#' random effects are used directly to characterize an outbreak.
#'
#' @param data A tibble containing the time series data, including columns 'y'
#' for observed values,'n' for population size, and other covariates of
#' interest.
#' @param formula A model formula for the fixed effects in the  hierarchical
#' model to fit to the data.
#' @param k An integer specifying the rolling window size employed for parameter
#' estimation.
#' @param sig_level The quantile from the random effects distribution used for
#' defining the for outbreak detection threshold, a numeric value between 0 and
#' 1.
#' @param exclude_past_outbreaks logical value indicating whether past outbreak
#' related observations should be excluded from future parameter estimation.
#' @param init_theta Initial values for model parameters in optimization.
#' @param lower Lower bounds for optimization parameters.
#' @param upper Upper bounds for optimization parameters.
#' @param method The optimization method to use, either "BFGS"  (default) or
#' "L-BFGS-B".
#'
#' @return A [tibble][tibble::tibble-package]-like 'aedseo' object containing:
#'   - 'window_data': A list of [tibble][tibble::tibble-package], each
#'   representing the data for this windowed parameter estimation.
#'   - 'reference_data': A list of [tibble][tibble::tibble-package], each
#'   representing the data for the reference time point.
#'   - 'phi': The dispersion parameter.
#'   - 'lambda': The estimated outbreak intensity.
#'   - 'u': The one-step ahead random effect.
#'   - 'u_probability': The probability of observing the one-step ahead
#'   random effect.
#'   - 'outbreak_alarm': Logical. Indicates if an outbreak is detected.
#'
#' @export
#'
#' @examples
#' # Create an example aedseo_tsd object
#' aeddo_data <- data.frame(
#'   time = as.Date(c(
#'     "2023-01-01",
#'     "2023-01-02",
#'     "2023-01-03",
#'     "2023-01-04",
#'     "2023-01-05",
#'     "2023-01-06"
#'   )),
#'   y = c(100, 120, 180, 110, 130, 140),
#'   n = 1
#' )
#'
#' # Supply a model formula
#' fixed_effects_formula <- y ~ 1
#'
#' # Choose a size for the rolling window
#' k <- 2
#' # ... and quantile for the threshold
#' sig_level <- 0.9
#'
#' # Employ the algorithm
#' aeddo_results <- aeddo(
#'   data = aeddo_data,
#'   formula = fixed_effects_formula,
#'   k = k,
#'   sig_level = sig_level,
#'   exclude_past_outbreaks = TRUE,
#'   init_theta = c(1, 0),
#'   lower = c(-Inf, 1e-6),
#'   upper = c(Inf, 1e2),
#'   method = "L-BFGS-B"
#' )
#' # Print the results
#' print(aeddo_results)
aeddo <- function(
    data = data.frame(),
    formula = formula(),
    k = integer(),
    sig_level = 0.95,
    exclude_past_outbreaks = TRUE,
    init_theta = numeric(),
    lower = numeric(),
    upper = numeric(),
    method = "BFGS") {
  # Assert function inputs
  check_aeddo_inputs(
    data,
    formula,
    k,
    sig_level,
    exclude_past_outbreaks,
    init_theta,
    lower,
    upper,
    method
  )

  # Count the number of observations
  n_observation <- dplyr::count(data) %>%
    purrr::pluck("n")

  # Prepare an empty tibble for results with class 'aedseo'.
  results <- tibble::new_tibble(
    x = tibble::tibble(
      window_data = list(),
      reference_data = list(),
      phi = numeric(),
      lambda = numeric(),
      u = numeric(),
      u_probability = numeric(),
      outbreak_alarm = logical()
    ),
    class = "aeddo",
    data = data,
    k = k
  )

  # Initilize past_outbreaks if we want to omit outbreak related observations
  if (exclude_past_outbreaks == TRUE) {
    past_outbreaks <- tibble::as_tibble(
      lapply(data, function(col) col[0])
    )
  }

  # Set starting guess for theta 
  theta <- init_theta # TODO: #17 Implement a method to infer the initial `theta`, `lower`, and `upper`

  # Loop over the observations to perform windowed estimation
  for (i in 1:(n_observation - k)) {
    # Extract data point for this estimation window
    window_data <- data %>%
      dplyr::filter(dplyr::row_number() %in% i:(i + k - 1))
    # ... and the reference data
    reference_data <- data %>%
      dplyr::filter(dplyr::row_number() == i + k)

    # Exclude past observations, if they were deemed an outbreak
    # Turned on by 'excludePastOutbreaks = TRUE'
    if (exclude_past_outbreaks == TRUE) {
      if (nrow(past_outbreaks) > 0) {
        window_data <- window_data %>%
          dplyr::setdiff(past_outbreaks)
      }
    }

    # Extract the observation at reference time point
    reference_y <- reference_data %>%
      purrr::pluck("y")
    # ... and the population size
    reference_n <- reference_data %>%
      purrr::pluck("n")

    # Construct a design matrix for the reference data
    reference_design_matrix <- stats::model.matrix(
      object = formula,
      data = reference_data
    )

    # Gather all arguments in a list for`do.call()`
    optimiser_arguments <- list(
      par = theta,
      data = window_data,
      formula = formula,
      fn = nll_poisson_gamma,
      method = method
    )

    if (method == "L-BFGS-B") {
      optimiser_arguments$lower <- lower
      optimiser_arguments$upper <- upper
    }

    # Obtain parameters with maximum likelihood estimation
    optimized_param <- do.call(what = "optim", args = optimiser_arguments)

    # Extract fixed effects parameters
    fixef_parameter <- optimized_param$par[seq_len(
      ncol(reference_design_matrix)
    )]
    # ... and dispersion parameter, phi
    phi <- exp(optimized_param$par[length(optimized_param$par)])

    # Establish one-step ahead lambda using reference observation
    lambda <- c(
      exp(reference_design_matrix %*% fixef_parameter - log(reference_n))
    )

    # Infer the one-step ahead random effect
    u <- (reference_y * phi + 1) / (lambda * phi + 1)

    # Compute probability this random effects is observed
    u_probability <- stats::pgamma(
      q = u,
      shape = 1 / phi,
      scale = phi
    )

    # Does the inferred one-step ahead random effect exceed significance level?
    outbreak_alarm <- u_probability >= sig_level

    # Gather results for return
    results <- results %>%
      tibble::add_row(
        window_data = list(window_data),
        reference_data = list(reference_data),
        phi = phi,
        lambda = lambda,
        u = u,
        u_probability = u_probability,
        outbreak_alarm = outbreak_alarm
      )

    if (exclude_past_outbreaks == TRUE && outbreak_alarm == TRUE) {
      past_outbreaks <- past_outbreaks %>%
        dplyr::bind_rows(reference_data)
    }

    # Set starting guess for theta given last round of optimization
    theta <- optimized_param$par * 0.7
  }

  return(results)
}
